Node2nix fails to install with nodejs-15 and nodejs-16

shell.nix
```
with import <nixpkgs> {};

let
    nodejs = nodejs-14_x;  # STILL WORKS
    # nodejs = nodejs-15_x;  # FAILS
    # nodejs = nodejs-16_x;  # FAILS
    nodePackages = pkgs.nodePackages.override {
      inherit nodejs;
    };

in

stdenv.mkDerivation rec {
  name = baseNameOf (builtins.getEnv "PWD");
  buildInputs = [
    nodejs
    nodePackages.node2nix
  ];
}
```

Failure for nodesjs-15 and nodejs-16 is identical.
```
nix-shell
...
rebuilt dependencies successfully
npm ERR! code ENOTCACHEDlTree buildDeps1ms
npm ERR! request to https://registry.npmjs.org/jsdoc failed: cache mode is 'only-if-cached' but no cached response available.

npm ERR! A complete log of this run can be found in:
npm ERR!     /build/.npm/_logs/2021-05-03T17_01_25_614Z-debug.log

builder for '/nix/store/ymrcv3axb8rfmvi1pk3yf03wbjgk505b-node_node2nix-1.9.0.drv' failed with exit code 1
error: build of '/nix/store/ymrcv3axb8rfmvi1pk3yf03wbjgk505b-node_node2nix-1.9.0.drv' failed
```

The current release still doesn’t work with newer Node.js. `node2nix` reconstructs a `package-lock.json` file with  `lockFileVersion: 1`; the newer `npm` tries to convert it to `lockFileVersion: 2`, and fails for lack of network access in the Nix sandbox. This is the error message originally reported in this issue.

```console
$ echo '["chalk"]' > node-packages.json

$ npx node2nix@1.11.0 --version
node2nix 1.11.0

$ npx node2nix@1.11.0 --nodejs-16 --input node-packages.json
info attempt registry request try #1 at 6:09:03 PM
http request GET https://registry.npmjs.org/chalk
http 200 https://registry.npmjs.org/chalk

$ nix-build
this derivation will be built:
  /nix/store/vxhbcl0nbz57yb2h8nasd95cryxglxaq-chalk-5.0.1.drv
building '/nix/store/vxhbcl0nbz57yb2h8nasd95cryxglxaq-chalk-5.0.1.drv'...
unpacking sources
patching sources
configuring
no configure script, doing nothing
building
installing
unpacking source archive /nix/store/dld45vw4jhv1scs7j6mz448ljwf41b1p-chalk-5.0.1.tgz
pinpointing versions of dependencies...
patching script interpreter paths in .
No package-lock.json file found, reconstructing...
rebuilt dependencies successfully
npm WARN old lockfile marborist:ctor Completed in 0ms
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
npm WARN old lockfile 
npm ERR! code ENOTCACHED
npm ERR! request to https://registry.npmjs.org/@types%2fnode failed: cache mode is 'only-if-cached' but no cached response is available.

npm ERR! A complete log of this run can be found in:
npm ERR!     /build/.npm/_logs/2022-04-23T01_09_41_970Z-debug-0.log

error: builder for '/nix/store/vxhbcl0nbz57yb2h8nasd95cryxglxaq-chalk-5.0.1.drv' failed with exit code 1;
       last 10 log lines:
       > npm WARN old lockfile so supplemental metadata must be fetched from the registry.
       > npm WARN old lockfile
       > npm WARN old lockfile This is a one-time fix-up, please be patient...
       > npm WARN old lockfile
       > npm ERR! code ENOTCACHED
       > npm ERR! request to https://registry.npmjs.org/@types%2fnode failed: cache mode is 'only-if-cached' but no cached response is available.
       > 
       > npm ERR! A complete log of this run can be found in:
       > npm ERR!     /build/.npm/_logs/2022-04-23T01_09_41_970Z-debug-0.log
       > 
       For full logs, run 'nix log /nix/store/vxhbcl0nbz57yb2h8nasd95cryxglxaq-chalk-5.0.1.drv'.
```

--- a/bin/node2nix.js
+++ b/bin/node2nix.js
@@ -28,7 +28,7 @@ var switches = [
     ['-18', '--nodejs-18', 'Provides all settings to generate expression for usage with Node.js 18.x (default is: nodejs-14_x)'],
     ['--supplement-input FILE', 'A supplement package JSON file that are passed as build inputs to all packages defined in the input JSON file'],
     ['--supplement-output FILE', 'Path to a Nix expression representing a supplementing set of Nix packages provided as inputs to a project (defaults to: supplement.nix)'],
-    ['--include-peer-dependencies', 'Specifies whether to include peer dependencies. In npm 2.x, this is the default. (false by default)'],
+    ['--include-peer-dependencies', 'Specifies whether to include peer dependencies. In npm 2.x, this is the default. (true by default for Node.js 16+)'],
     ['--no-flatten', 'Simulate pre-npm 3.x isolated dependency structure. (false by default)'],
     ['--pkg-name NAME', 'Specifies the name of the Node.js package to use from Nixpkgs (defaults to: nodejs)'],
     ['--registry URL', 'URL referring to the NPM packages registry. It defaults to the official NPM one, but can be overridden to support private registries'],
@@ -47,7 +47,7 @@ var parser = new optparse.OptionParser(switches);
 var help = false;
 var version = false;
 var production = true;
-var includePeerDependencies = false;
+var includePeerDependencies = true;
 var flatten = true;
 var inputJSON = "package.json";
 var outputNix = "node-packages.nix";
@@ -118,61 +118,71 @@ parser.on('development', function(arg, value) {
 parser.on('nodejs-4', function(arg, value) {
     flatten = false;
     nodePackage = "nodejs-4_x";
-    byPassCache = false;
+    bypassCache = false;
+    includePeerDependencies = false;
 });
 
 parser.on('nodejs-6', function(arg, value) {
     flatten = true;
     nodePackage = "nodejs-6_x";
-    byPassCache = false;
+    bypassCache = false;
+    includePeerDependencies = false;
 });
 
 parser.on('nodejs-8', function(arg, value) {
     flatten = true;
     nodePackage = "nodejs-8_x";
     bypassCache = true;
+    includePeerDependencies = false;
 });
 
 parser.on('nodejs-10', function(arg, value) {
     flatten = true;
     nodePackage = "nodejs-10_x";
     bypassCache = true;
+    includePeerDependencies = false;
 });
 
 parser.on('nodejs-12', function(arg, value) {
     flatten = true;
     nodePackage = "nodejs-12_x";
     bypassCache = true;
+    includePeerDependencies = false;
 });
 
 parser.on('nodejs-13', function(arg, value) {
     flatten = true;
     nodePackage = "nodejs-13_x";
     bypassCache = true;
+    includePeerDependencies = false;
 });
 
 parser.on('nodejs-14', function(arg, value) {
     flatten = true;
     nodePackage = "nodejs-14_x";
     bypassCache = true;
+    includePeerDependencies = false;
 });
 
 parser.on('nodejs-16', function(arg, value) {
     flatten = true;
     nodePackage = "nodejs-16_x";
     bypassCache = true;
+    includePeerDependencies = true;
 });
 
 parser.on('nodejs-17', function(arg, value) {
     flatten = true;
     nodePackage = "nodejs-17_x";
     bypassCache = true;
+    includePeerDependencies = true;
 });
 
 parser.on('nodejs-18', function(arg, value) {
     flatten = true;
     nodePackage = "nodejs-18_x";
     bypassCache = true;
+    includePeerDependencies = true;
 });
 
 parser.on('include-peer-dependencies', function(arg, value) {

--- a/lib/Package.js
+++ b/lib/Package.js
@@ -196,7 +196,7 @@ Package.prototype.resolveDependenciesAndSources = function(callback) {
 
         function(callback) {
             if(self.deploymentConfig.includePeerDependencies) {
-                /* Bundle the peer dependencies, if applicable */
+                /* Bundle the required peer dependencies, if applicable */
                 self.bundleDependencies(resolvedDependencies, self.source.config.peerDependencies, callback);
             } else {
                 callback();

--- a/nix/node-env.nix
+++ b/nix/node-env.nix
@@ -165,7 +165,11 @@ let
           if(process.argv[2] == "development") {
               replaceDependencies(packageObj.devDependencies);
           }
+          else {
+              packageObj.devDependencies = {};
+          }
           replaceDependencies(packageObj.optionalDependencies);
+          replaceDependencies(packageObj.peerDependencies);
 
           /* Write the fixed package.json file */
           fs.writeFileSync("package.json", JSON.stringify(packageObj, null, 2));
@@ -270,7 +274,7 @@ let
 
   # Reconstructs a package-lock file from the node_modules/ folder structure and package.json files with dummy sha1 hashes
   reconstructPackageLock = writeTextFile {
-    name = "addintegrityfields.js";
+    name = "reconstructpackagelock.js";
     text = ''
       var fs = require('fs');
       var path = require('path');
@@ -280,25 +284,43 @@ let
       var lockObj = {
           name: packageObj.name,
           version: packageObj.version,
-          lockfileVersion: 1,
+          lockfileVersion: 2,
           requires: true,
+          packages: {
+              "": {
+                  name: packageObj.name,
+                  version: packageObj.version,
+                  license: packageObj.license,
+                  bin: packageObj.bin,
+                  dependencies: packageObj.dependencies,
+                  engines: packageObj.engines,
+                  optionalDependencies: packageObj.optionalDependencies
+              }
+          },
           dependencies: {}
       };
 
-      function augmentPackageJSON(filePath, dependencies) {
+      function augmentPackageJSON(filePath, packages, dependencies) {
           var packageJSON = path.join(filePath, "package.json");
           if(fs.existsSync(packageJSON)) {
               var packageObj = JSON.parse(fs.readFileSync(packageJSON));
+              packages[filePath] = {
+                  version: packageObj.version,
+                  integrity: "sha1-000000000000000000000000000=",
+                  dependencies: packageObj.dependencies,
+                  engines: packageObj.engines,
+                  optionalDependencies: packageObj.optionalDependencies
+              };
               dependencies[packageObj.name] = {
                   version: packageObj.version,
                   integrity: "sha1-000000000000000000000000000=",
                   dependencies: {}
               };
-              processDependencies(path.join(filePath, "node_modules"), dependencies[packageObj.name].dependencies);
+              processDependencies(path.join(filePath, "node_modules"), packages, dependencies[packageObj.name].dependencies);
           }
       }
 
-      function processDependencies(dir, dependencies) {
+      function processDependencies(dir, packages, dependencies) {
           if(fs.existsSync(dir)) {
               var files = fs.readdirSync(dir);
 
@@ -314,20 +336,81 @@ let
                           pkgFiles.forEach(function(entry) {
                               if(stats.isDirectory()) {
                                   var pkgFilePath = path.join(filePath, entry);
-                                  augmentPackageJSON(pkgFilePath, dependencies);
+                                  augmentPackageJSON(pkgFilePath, packages, dependencies);
                               }
                           });
                       } else {
-                          augmentPackageJSON(filePath, dependencies);
+                          augmentPackageJSON(filePath, packages, dependencies);
                       }
                   }
               });
           }
       }
 
-      processDependencies("node_modules", lockObj.dependencies);
+      processDependencies("node_modules", lockObj.packages, lockObj.dependencies);
 
       fs.writeFileSync("package-lock.json", JSON.stringify(lockObj, null, 2));
+    '';
+  };
+
+  # Script that links bins defined in package.json to the node_modules bin directory
+  # NPM does not do this for top-level packages itself anymore as of v7
+  linkBinsScript = writeTextFile {
+    name = "linkbins.js";
+    text = ''
+      var fs = require('fs');
+      var path = require('path');
+
+      var packageObj = JSON.parse(fs.readFileSync("package.json"));
+
+      var nodeModules = Array(packageObj.name.split("/").length).fill("..").join(path.sep);
+
+      if(packageObj.bin !== undefined) {
+          fs.mkdirSync(path.join(nodeModules, ".bin"))
+
+          if(typeof packageObj.bin == "object") {
+              Object.keys(packageObj.bin).forEach(function(exe) {
+                  if(fs.existsSync(packageObj.bin[exe])) {
+                      console.log("linking bin '" + exe + "'");
+                      fs.symlinkSync(
+                          path.join("..", packageObj.name, packageObj.bin[exe]),
+                          path.join(nodeModules, ".bin", exe)
+                      );
+                  }
+                  else {
+                      console.log("skipping non-existent bin '" + exe + "'");
+                  }
+              })
+          }
+          else {
+              if(fs.existsSync(packageObj.bin)) {
+                  console.log("linking bin '" + packageObj.bin + "'");
+                  fs.symlinkSync(
+                      path.join("..", packageObj.name, packageObj.bin),
+                      path.join(nodeModules, ".bin", packageObj.name.split("/").pop())
+                  );
+              }
+              else {
+                  console.log("skipping non-existent bin '" + packageObj.bin + "'");
+              }
+          }
+      }
+      else if(packageObj.directories !== undefined && packageObj.directories.bin !== undefined) {
+          fs.mkdirSync(path.join(nodeModules, ".bin"))
+
+          fs.readdirSync(packageObj.directories.bin).forEach(function(exe) {
+              if(fs.existsSync(path.join(packageObj.directories.bin, exe))) {
+                  console.log("linking bin '" + exe + "'");
+                  fs.symlinkSync(
+                      path.join("..", packageObj.name, packageObj.directories.bin, exe),
+                      path.join(nodeModules, ".bin", exe)
+                  );
+              }
+              else {
+                  console.log("skipping non-existent bin '" + exe + "'");
+              }
+          })
+      }
     '';
   };
 
@@ -377,13 +460,18 @@ let
 
         npm ${forceOfflineFlag} --nodedir=${nodeSources} ${npmFlags} ${lib.optionalString production "--production"} rebuild
 
+        runHook postRebuild
+
         if [ "''${dontNpmInstall-}" != "1" ]
         then
             # NPM tries to download packages even when they already exist if npm-shrinkwrap is used.
             rm -f npm-shrinkwrap.json
 
-            npm ${forceOfflineFlag} --nodedir=${nodeSources} ${npmFlags} ${lib.optionalString production "--production"} install
+            npm ${forceOfflineFlag} --nodedir=${nodeSources} --no-bin-links --ignore-scripts ${npmFlags} ${lib.optionalString production "--production"} install
         fi
+
+        # Link executables defined in package.json
+        node ${linkBinsScript}
     '';
 
   # Builds and composes an NPM package including all its dependencies

